cpp/src/arrow/util/bit_stream_utils.h

```diff
@@ -204,8 +204,8 @@ class BitReader {
 
 inline bool BitWriter::PutValue(uint64_t v, int num_bits) {
   // TODO: revisit this limit if necessary (can be raised to 64 by fixing some edge cases)
-  DCHECK_LE(num_bits, 32);
-  DCHECK_EQ(v >> num_bits, 0) << "v = " << v << ", num_bits = " << num_bits;
+  DCHECK_LE(num_bits, 64);
+  //  DCHECK_EQ(v >> num_bits, 0) << "v = " << v << ", num_bits = " << num_bits;
```

Comment:
This errors for `int64` (I suppose it wraps around?), but I we don't want to disable the check. @pitrou shall I make a version of `PutValue`without the check?

Reply:
Well, what error does it display? It should show you the precise values...

Reply:
Or you can try avoiding the check in case num_bits is 64.
```suggestion
  if (num_bits < 64) {
    DCHECK_EQ(v >> num_bits, 0) << "v = " << v << ", num_bits = " << num_bits;
  }
```

Reply:
Well, the check ensures that `v` fits in `num_bits`.

Reply:
However, the `v >> (num_bits - bit_offset_)` operation below might be undefined if `num_bits - bit_offset_` is 64.
So I would change:
```c++
    buffered_values_ = v >> (num_bits - bit_offset_);
```
into:
```c++
    buffered_values_ = (num_bits - bit_offset_ == 64) ? 0 : (v >> (num_bits - bit_offset_));
```

Reply:
This seems to fix the overflow issue for `int64`!